Enforce response body size limits in payjoin-cli

GideonBature opened the pull request:
This PR follows up on #770 and addresses the remaining improvements in [comment](https://github.com/payjoin/rust-payjoin/pull/770#issuecomment-3000712651). It tightens content-length handling on the sender and enforces body size limits in `payjoin-cli`, so a malicious counterparty cannot trigger unbounded memory allocation.

### Summary of changes

1. **Sender-side validation**
   - Clarifies the sender's `ContentTooLarge` error so it no longer leaks the internal `MAX_CONTENT_LENGTH` value, and rewrites the v1 response tests to assert the content-length boundary (under, equal to, and over the limit).

2. **`payjoin-cli` receiver**
   - Adds a size limit when reading the request body via a new `read_limited_body` helper (streaming with `into_data_stream()`), to avoid unbounded memory allocations from a potentially malicious counterparty.

3. **`payjoin-cli` sender**
   - Enforces the same limit when handling response bodies, streaming via `bytes_stream()` from `reqwest` and rejecting upfront when the `Content-Length` header exceeds the cap.

Ref: #770
Closes: #756

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -88,7 +93,19 @@ impl AppTrait for App {
+            .and_then(|val| val.to_str().ok())
+            .and_then(|s| s.parse::<usize>().ok());
+
+        if content_length.unwrap() > MAX_CONTENT_LENGTH {
```
Noted...

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -88,7 +93,19 @@ impl AppTrait for App {
+            return Err(anyhow!("Response body is too large: {:?} bytes", content_length));
+        }
+
+        let psbt = ctx.process_response(&response.bytes().await?, content_length).map_err(|e| {
```
Noted!

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -276,9 +293,24 @@ impl App {
+        };
+
         let query_string = parts.uri.query().unwrap_or("");
         let body = body.collect().await.map_err(|e| Implementation(e.into()))?.to_bytes();
```
Noted!

GideonBature replied:
Noted!

GideonBature commented on payjoin/src/send/v1.rs:
```diff
@@ -278,9 +278,19 @@ impl V1Context {
+    pub fn process_response(
+        self,
+        response: &[u8],
+        content_length: Option<usize>,
```
noted!

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -88,7 +93,19 @@ impl AppTrait for App {
         );
-        let psbt = ctx.process_response(&response.bytes().await?).map_err(|e| {
+
+        let content_length = response
```
Thank you for this.

nothingmuch reviewed (commented):
This is looking great. Thank you!

The new dependencies are fine IMO, because they are already in our dependency chain if I'm not mistaken, and also because payjoin-cli can be more flexible about new dependencies than the payjoin crate.

The buffering code is very clear. I think a `Vec<u8>` can replace `Bytes`, since they are more or less the same on the inside, the main difference is `Bytes` support cheap `clone()`, which we don't need, for the cost of slightly more indirection.

Instead of duplicating it in the sender and receiver, a utility function can be added to payjoin-cli's `app` module.

Finally, as a nitpick, I find additional whitespace to be helpful when reading, but please try to commit such trivial changes in their own commits, that makes it easier to review commit by commit, github's files view uses lexicographical order for files and combines all the changes of all commits, both of which make it harder to spot important changes in between trivial ones.

GideonBature commented:
I implemented the suggestions you gave, however, for the utility function, putting it in `mod.rs` of `app`, I am encountering a challenge where I am unable to access the `hyper` dependencies, I really don't know why. And I will be needing it, since I will need to make use of `Bytes` from it for the utility function.

nothingmuch reviewed (changes requested):
I didn't quite understand your last comment about hyper & bytes, but i think the `read_limited_body` function is in the right place?

similar limiting for `app/v2.rs` can be added in a separate PR, with a fixed response size, and reusing this function, but moving it to the top level `mod.rs` should be straightforward (just making it `pub(crate)` and bringing in the necessary imports, namely `hyper::body::Bytes`)

anyway this is getting close, i only have some fairly small cleanup suggestions

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -357,3 +396,21 @@ fn try_contributing_inputs(
+
+    while let Some(chunk) = stream.next().await {
+        let chunk = chunk.map_err(|e| anyhow!("Error reading body chunk: {}", e))?;
+        if body.len() + chunk.len() > max_len {
+            return Err(anyhow!("Body exceeds maximum allowed size of {max_len} bytes"));
```
Noted!

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -276,9 +296,28 @@ impl App {
+            ));
+        }
+
+        let body = read_limited_body(body.into_data_stream(), MAX_CONTENT_LENGTH)
```
Noted!

GideonBature commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -276,9 +296,28 @@ impl App {
         let query_string = parts.uri.query().unwrap_or("");
-        let body = body.collect().await.map_err(|e| Implementation(e.into()))?.to_bytes();
-        let proposal = UncheckedProposal::from_request(&body, query_string, headers)?;
+        let body = Bytes::from(body);
+        let proposal = UncheckedProposal::from_request(&body, query_string, headers.clone())?;
```
Noted!

GideonBature commented:
> I didn't quite understand your last comment about hyper & bytes, but i think the `read_limited_body` function is in the right place?
> 
> similar limiting for `app/v2.rs` can be added in a separate PR, with a fixed response size, and reusing this function, but moving it to the top level `mod.rs` should be straightforward (just making it `pub(crate)` and bringing in the necessary imports, namely `hyper::body::Bytes`)
> 
> anyway this is getting close, i only have some fairly small cleanup suggestions

Moving the function to `mod.rs`, I always have issue bringing in the `hyper` dependency. This is the error I always get, despite `hyper` being present in the `Cargo.toml` file.

DanGould commented on payjoin-cli/Cargo.toml:
```diff
@@ -24,6 +24,8 @@ native-certs = ["reqwest/rustls-tls-native-roots"]
 _danger-local-https = ["rcgen", "reqwest/rustls-tls", "rustls", "hyper-rustls", "payjoin/_danger-local-https", "tokio-rustls"]
 v1 = ["payjoin/v1","hyper", "hyper-util", "http-body-util"]
 v2 = ["payjoin/v2", "payjoin/io"]
+hyper = ["dep:hyper"]
+http-body-util = ["dep:http-body-util"]
```
It seems that you've added hyper as an optional dependency which is only enabled when the "hyper" feature is enabled, which it only is when using the `v1` feature. If you're building `payjoin-cli` with the default feature set that does not enable `v1` or the `hyper` feature you've defined here. Rather than making specific features for hyper and http-body-util, just enable them in the appropriate feature combination by including them in the `[...]` lists or making them default (by removing optional = true) if they are used with every feature combination

GideonBature commented:
This PR addresses the hyper crate issue and also all the suggestions given. Thank you.

GideonBature commented:
Resolved merge conflicts.

DanGould commented on payjoin/src/core/send/mod.rs:
```diff
@@ -718,10 +718,6 @@ impl V1Context {
     /// valid you will get appropriate PSBT that you should sign and broadcast.
     #[inline]
     pub fn process_response(self, response: &[u8]) -> Result<Psbt, ResponseError> {
-        if response.len() > MAX_CONTENT_LENGTH {
-            return Err(ResponseError::from(InternalValidationError::ContentTooLarge));
-        }
```
@nothingmuch what's the rationale for making the client do this? Just because we're not taking a `Read` anymore it doesn't make sense to do here?

DanGould commented on payjoin-cli/src/app/mod.rs:
```diff
@@ -80,3 +82,22 @@ async fn handle_interrupt(tx: watch::Sender<()>) {
     let _ = tx.send(());
 }
+
+#[allow(dead_code)]
```
The warning that required you to add this demonstrates there was probably something wrong with where this lives. It's only used in v1 so it should live in v1, which would let you get around this warning.

GideonBature replied:
Noted!

DanGould commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -4,6 +4,8 @@ use std::str::FromStr;
 use std::sync::Arc;
 
 use anyhow::{anyhow, Context, Result};
+#[allow(unused_imports)]
+use futures::{Stream, StreamExt};
```
Just delete the import. No need to allow it.

GideonBature replied:
Noted!

DanGould commented on payjoin/src/core/send/v1.rs:
```diff
@@ -303,6 +306,7 @@ mod test {
             "message": "This version of payjoin is not supported."
         })
         .to_string();
+
```
Please leave unrelated whitespace changes like this for their own commits. They make me wonder why you were in this file when I review.

GideonBature replied:
Noted! Seems it has to do with the configuration of my code editor!

DanGould replied:
These unrelated whitespace changes are still here in many places in this file. May you please remove them?

DanGould commented on payjoin/src/core/send/v1.rs:
```diff
@@ -371,35 +378,4 @@ mod test {
-        assert_eq!(MAX_CONTENT_LENGTH, 4_000_000 * 4 / 3);
-    }
-
-    #[test]
-    fn test_non_witness_input_weight_const() {
-        assert_eq!(NON_WITNESS_INPUT_WEIGHT, bitcoin::Weight::from_wu(160));
-    }
```
I don't think this test was meant to be removed

DanGould commented on payjoin/tests/integration.rs:
```diff
@@ -99,7 +99,8 @@ mod integration {
             // Inside the Sender:
             // Sender checks, signs, finalizes, extracts, and broadcasts
-            let checked_payjoin_proposal_psbt = ctx.process_response(response.as_bytes())?;
+            let response_body = response.as_bytes();
+            let checked_payjoin_proposal_psbt = ctx.process_response(response_body)?;
```
Please remove these changes, they're not related to the purpose of this PR.

GideonBature replied:
Done, although seems why I had to break it down is the fact that clippy was complaining.

DanGould reviewed (changes requested):
Thanks for following up. The core behavior here seems right now 😎

 but there appear to be 2 problems

1. There are extraneous changes not related to the goal of the PR. I left specific comments where, please remove them
2. This PR removes an unrelated test. Please restore it.


One of my comments related to feature gating was also not addressed, so I went ahead and addressed it so that you can see the example of what needed to change for the future.

I also went ahead and combined an abstraction I saw that remained at both places you implemented read_limited_body.

GideonBature commented on .Cargo.lock.flock:
Done! Thank you!

nothingmuch reviewed (commented):
I appear to have compeltely forgotten about this PR, I'm really sorry about this...

there appear to be unrelated changes, likely the result of an incorrect rebase? the changes having to do with the length enforcement seem correct to me but i'm not sure why some things like `V1Context` are there

GideonBature commented:
Oh, I will do a check and fix that the push for your review.

nothingmuch reviewed (commented):
utACK, but commit needs cleanup:

1. lcov.info shouldn't be added, should probably be in gitignore but that's a separate issue

2. i don't understand why rcgen was added to Cargo.toml, i think only `stream` feature to `reqwest` was intended in this PR?

GideonBature commented:
> utACK, but commit needs cleanup:
> 
> 1. lcov.info shouldn't be added, should probably be in gitignore but that's a separate issue
> 2. i don't understand why rcgen was added to Cargo.toml, i think only `stream` feature to `reqwest` was intended in this PR?

Yes, I think I made a mixup, let me remove it. Then create an issue to add `Icov.info` to `.gitignore`.

nothingmuch commented on payjoin/src/core/send/v1.rs:
```diff
@@ -349,6 +351,16 @@ mod test {
+    #[test]
+    fn test_non_witness_input_weight_const() {
+        assert_eq!(NON_WITNESS_INPUT_WEIGHT, bitcoin::Weight::from_wu(160));
+    }
```
i don't understand the purpose of these tests

nothingmuch commented on payjoin/src/core/send/v1.rs:
```diff
@@ -490,12 +506,12 @@ mod test {
-        let mut invalid_utf8_padding = PAYJOIN_PROPOSAL.as_bytes().to_vec();
-        invalid_utf8_padding
-            .extend(std::iter::repeat_n(0x00, MAX_CONTENT_LENGTH - invalid_utf8_padding.len()));
+        let mut invalid_utf8 = PAYJOIN_PROPOSAL.as_bytes().to_vec();
```
this variable renaming seems unrelated

nothingmuch commented on payjoin/src/core/send/v1.rs:
```diff
@@ -29,8 +29,10 @@ use error::BuildSenderError;
 pub use crate::output_substitution::OutputSubstitution;
-use crate::uri::v1::PjParam;
-use crate::{PjUri, Request, MAX_CONTENT_LENGTH};
+#[allow(unused_imports)]
```
why?

nothingmuch commented on payjoin/src/core/send/mod.rs:
```diff
@@ -706,6 +710,62 @@ mod test {
         })
     }
 
+    #[test]
```
what's the reason these tests are here and not in v1.rs?

nothingmuch commented on payjoin/src/core/send/v1.rs:
```diff
@@ -198,7 +200,7 @@ impl Sender {
 #[derive(Debug, Clone)]
 pub struct V1Context {
-    psbt_context: PsbtContext,
+    pub(crate) psbt_context: PsbtContext,
```
seems like this pub(crate) because the tests that use it are in mod.rs but i think if they were moved to this file that would no longer be necessary?

nothingmuch commented on payjoin/src/core/send/v1.rs:
```diff
@@ -69,7 +71,7 @@ impl SenderBuilder {
     ) -> Self {
         Self {
-            endpoint: pj_param.endpoint(),
+            endpoint: pj_param.endpoint_url(),
```
this seems to confuse the payjoin crate's `Url` type which is used to avoid depending on `url::Url`, which `endpoint_url` returns.

@benalleng what do you think about removing `endpoint_url`? any reason that wasn't done in #1377 i'm missing?

benalleng replied:
Thinking back I kept it probably just because I didn't feel the need to change it to keep the behavior consistent where I could.  Though perhaps `endpoint()` vs `endpoint_url()` is too easily mistaken.

DanGould commented:
@GideonBature do you have plans to continue work here? If not, we can close for now and revisit when there's more capacity for it. Icov.info is now be gitignored.

GideonBature commented:
Yes @DanGould I will finish it up... Thank you for reaching out.

DanGould commented:
Thanks for the fast turnaround on this @GideonBature. Please rebase the history so that formatting changes are squashed into the commit that introduces the code and so that the rules are followed according to [the contributing guidelines](https://github.com/payjoin/rust-payjoin?tab=contributing-ov-file#commits), the intermediate commit 6bd1c78 failing in CI suggests that these steps aren't done yet. Seems like there should be two commits passing tests in this PR according to those rules. There are commit message rules there, too. Once those are done, happy to review.

GideonBature commented:
> Thanks for the fast turnaround on this @GideonBature. Please rebase the history so that formatting changes are squashed into the commit that introduces the code and so that the rules are followed according to [the contributing guidelines](https://github.com/payjoin/rust-payjoin?tab=contributing-ov-file#commits), the intermediate commit [6bd1c78](https://github.com/payjoin/rust-payjoin/commit/6bd1c78ff654a72a0f77a7fba3a7acdc9cf5fd37) failing in CI suggests that these steps aren't done yet. Seems like there should be two commits passing tests in this PR according to those rules. There are commit message rules there, too. Once those are done, happy to review.

Noted. I will do that.

GideonBature committed:
Enforce response body size limits in payjoin-cli

Reading a streamed HTTP response into memory without an upper bound
lets a malicious counterparty exhaust memory and crash the peer's
payjoin-cli. Add a read_limited_body helper that aborts once the
accumulated body exceeds the expected length, and use it on both the
cli sender and receiver paths. This requires reqwest's "stream"
feature and the futures crate.

On the library side, the sender's ContentTooLarge message no longer
leaks the internal MAX_CONTENT_LENGTH value, and the v1 response
tests are rewritten to assert the content-length boundary directly
(under, equal to, and over the limit).

DanGould commented:
@GideonBature Thank you for following up.  I took a peek at this when it was just posted but was confused because the commit message rules don't seem to be addressed. Since this now needs to be rebased, may you please make sure the commit message follows the rules in CONTRIBUTING.md including rationale for the changes if you would be so kind as to rebase?

GideonBature commented:
> @GideonBature Thank you for following up. I took a peek at this when it was just posted but was confused because the commit message rules don't seem to be addressed. Since this now needs to be rebased, may you please make sure the commit message follows the rules in CONTRIBUTING.md including rationale for the changes if you would be so kind as to rebase?

Oh my bad. I will resolve this... Thank you...

DanGould commented on payjoin-cli/src/app/mod.rs:
```diff
@@ -103,3 +109,22 @@ async fn handle_interrupt(tx: watch::Sender<()>) {
     let _ = tx.send(());
 }
+
+#[cfg(feature = "v1")]
+pub async fn read_limited_body<S, E>(mut stream: S, expected_len: usize) -> Result<Vec<u8>>
+where
+    S: Stream<Item = Result<Bytes, E>> + Unpin,
+    E: std::error::Error + Send + Sync + 'static,
+{
+    let mut body = Vec::with_capacity(expected_len);
+
+    while let Some(chunk) = stream.next().await {
+        let chunk = chunk.map_err(|e| anyhow!("Error reading body chunk: {}", e))?;
+        if body.len() + chunk.len() > expected_len {
+            return Err(anyhow!("Body exceeds expected size of {expected_len} bytes"));
+        }
+        body.extend_from_slice(&chunk);
+    }
+
+    Ok(body)
+}
```
Because this is only called in `payjoin-cli/src/app/v1.rs` it ought to have dependencies declared there and be defined there. This'll get rid of all the feature gates in this file, too.

DanGould commented on payjoin-cli/src/app/v2/mod.rs:
```diff
@@ -172,7 +172,7 @@ impl AppTrait for App {
                 let (req, ctx) = payjoin::send::v1::SenderBuilder::from_parts(
                     psbt,
-                    pj_param,
+                    &PjParam::V1(pj_param.clone()),
```
This function's type signature does not change. Why then introduce this change?

DanGould commented on payjoin-cli/Cargo.toml:
```diff
@@ -42,6 +43,7 @@ r2d2_sqlite = "0.22.0"
 reqwest = { version = "0.12.23", default-features = false, features = [
   "json",
   "rustls-tls",
+  "stream",
```
This feature seems to only be used in v1, so rather than defining it here you should be able to define it in `v1 = [ ... ]`, above.

DanGould commented on payjoin-cli/Cargo.toml:
```diff
@@ -21,7 +21,7 @@ path = "src/main.rs"
 native-certs = ["reqwest/rustls-tls-native-roots"]
 _manual-tls = ["reqwest/rustls-tls", "payjoin/_manual-tls", "tokio-rustls"]
-v1 = ["payjoin/v1", "hyper", "hyper-util", "http-body-util"]
+v1 = ["payjoin/v1", "futures", "hyper", "hyper-util", "http-body-util"]
```
```suggestion
v1 = ["payjoin/v1", "futures", "hyper", "hyper-util", "http-body-util", "reqwest/stream"]
```

DanGould commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -71,36 +76,33 @@ impl AppTrait for App {
-        let psbt = match ctx.process_response(&response.bytes().await?) {
-            Ok(psbt) => psbt,
-            Err(e) => {
-                tracing::error!("Error processing response: {e:?}");
-                println!("Payjoin failed. To broadcast the fallback transaction, run:");
-                println!(
-                    "  bitcoin-cli -rpcwallet=<wallet> sendrawtransaction {:#}",
-                    serialize_hex(&fallback_tx)
-                );
-                return Err(anyhow!("Failed to process response {e}"));
-            }
-        };
+            .with_context(|| "HTTP request failed")?;
+        println!("Sent fallback transaction txid: {}", fallback_tx.compute_txid());
+        println!("Sent fallback transaction hex: {:#}", serialize_hex(&fallback_tx));
+
+        let expected_length = response
+            .headers()
+            .get("Content-Length")
+            .and_then(|val| val.to_str().ok())
+            .and_then(|s| s.parse::<usize>().ok())
+            .unwrap_or(MAX_CONTENT_LENGTH);
+
+        if expected_length > MAX_CONTENT_LENGTH {
+            return Err(anyhow!("Response body is too large: {} bytes", expected_length));
+        }
+
+        let body = read_limited_body(response.bytes_stream(), MAX_CONTENT_LENGTH).await?;
+
+        let psbt = ctx.process_response(&body).map_err(|e| {
+            tracing::debug!("Error processing response: {e:?}");
+            anyhow!("Failed to process response {e}")
+        })?;
 
         self.process_pj_response(psbt)?;
```
The println instructions may have changed since your first attempt at this. Please sure the println/error handling behavior doesn't change except for what you've introduced with Content-Length here. I imagine the lines like this

```
println!("Payjoin failed. To broadcast the fallback transaction, run:");
                println!(
                    "  bitcoin-cli -rpcwallet=<wallet> sendrawtransaction {:#}",
                    serialize_hex(&fallback_tx)
                );
```

can and should stay.

DanGould commented on payjoin/src/core/send/error.rs:
```diff
@@ -120,10 +120,7 @@ impl fmt::Display for ValidationError {
-                use crate::MAX_CONTENT_LENGTH;
-                write!(f, "content is larger than {MAX_CONTENT_LENGTH} bytes")
-            }
+            ContentTooLarge => write!(f, "The response body is too large"),
```
Maybe I'm missing context. What's the rationale for this change?

DanGould reviewed (changes requested):
Thank you for following up with the CONTRIBUTING.md prerequisites. There are a few more changes I'd like to request to get this ready. I tried to be thorough and believe if all of these are addressed this should be ready to go. tests look good.
